### PR TITLE
add a Deep Residual Network example for cifar10

### PR DESCRIPTION
**Pull request by freesouls**
Hi, I have add a Deep Residual Network example for cifar10

#### Commands & setups:
- 
set following parameters in example/image-classification/train_model.py 
momentum = 0.9, wd = 0.0001, initializer = mx.init.Xavier(rnd_type="gaussian", factor_type="in", magnitude=2.0)
- 
set n=3(3 for 20 layers, n=9 for 56 layers) in the get_symbol function in example/image-classification/symbol_resnet-28-small.py

- first train the network with lr=0.1 for 80 epochs
```
python example/image-classification/train_cifar10.py --network resnet-28-small --num-examples 50000 --lr 0.1 --num-epochs 80 --model-prefix cifar10/resnet 
```
- second train the network with lr=0.01 from epoch 81 to epoch 120, with lr=0.001 from epoch 121 to epoch 160
```
python example/image-classification/train_cifar10.py --network resnet-28-small --num-examples 50000 --model-prefix cifar10/resnet --load-epoch 80 --lr 0.01 --lr-factor 0.1 --lr-factor-epoch 40 --num-epochs 200 
```
in the paper, he train cifar10 for 160 epoch, I set num-epoch to 200 because I want to see whether it is usefull when set lr=0.0001

since it needs 160 epochs, please be patient

#### Accuracy:
- for 20 layers resnet, accuracy=0.905+
- for 32 layers resnet, accuracy=0.908+
- for 56 layers resnet, accuracy=0.915+

#### Differences to the paper on cifar10 network setup
1. in the paper, the author use identity shortcut when dealing with increasing dimensions, while I use 1x1 convolutions to deal with it
2. in the paper, 4 pixels are padded on each side and a 32x32 crop is randomly sampled from the padded image, while I use the dataset provided by mxnet, so the input is 28x28, as a results for 3 different kinds of 2n layers output map sizes are 28x28, 14x14, 7x7, instead of 32x32, 16x16, 8x8 in the paper.

the above two reason might answer why the accuracy is a bit lower than the paper, I suppose.
Off course, there might be other reasons (for example the true network architecture may be different from my script, since my script is just my understanding of the paper), if you find out, please tell me, declanxu@gmail.com or declanxu@126.com, thanks

@winstywang @tqchen @antinucleon

### CODE REVIEW

**Comment by freesouls** [comment]
there is a typo in train_cifar10.py, I think num-examples are 50000, If you want to reproduce my work, you might use num-examples 50000 as well, though generally speaking, it does not matter whether num-exmaples is 50000 or 60000.

**Comment by freesouls** [comment]
one more thing, I use batch-size of 128, and train the models by using one GPU. :)

**Comment by piiswrong** [comment]
Looks good to me. Could you add your net's results and the original paper's results in the beginning of your script?

**Comment by freesouls** [comment]
@piiswrong , added to the script
accuracy:
- for 20 layers resnet, accuracy=0.905+, 0.9125 in the paper
- for 32 layers resnet, accuracy=0.908+, 0.9239 in the paper
- for 56 layers resnet, accuracy=0.915+, 0.9303 in the paper

since the setups of the paper and my script are different, I could only say it does obey the rule: the deeper, the better

File: example/image-classification/symbol_resnet-28-small.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+        # adopt project method in the paper when dimension increased
+        project_data = conv_factory(data=data, num_filter=num_filter, kernel=(1,1), stride=(2,2), pad=(0,0), conv_type=1)
+        new_data = project_data + conv2
+        act = mx.symbol.Activation(data=new_data, act_type='relu')
+        return act
+
+def residual_net(data, n):
+    #fisrt 2n layers
+    for i in xrange(0, n):
```
Comment (antinucleon) [inline example/image-classification/symbol_resnet-28-small.py]: Change to ```range``` for python3

File: example/image-classification/symbol_resnet-28-small.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        data = residual_factory(data=data, num_filter=16, dim_match=True)
+    
+    #second 2n layers
+    for i in xrange(0, n):
```
Comment (antinucleon) [inline example/image-classification/symbol_resnet-28-small.py]: Change to ```range``` for python3

File: example/image-classification/symbol_resnet-28-small.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        if i==0:
+            data = residual_factory(data=data, num_filter=32, dim_match=False)
+        else:
+            data = residual_factory(data=data, num_filter=32, dim_match=True)
+    
+    #third 2n layers
+    for i in xrange(0, n):
```
Comment (antinucleon) [inline example/image-classification/symbol_resnet-28-small.py]: Change to ```range``` for python3

**Comment by antinucleon** [comment]
Please fix the python3 convention, then squash the commits: https://mxnet.readthedocs.org/en/latest/contribute.html#how-to-combine-multiple-commits-into-one

File: example/image-classification/symbol_resnet-28-small.py (hunk 3 above)
Comment (freesouls) [inline example/image-classification/symbol_resnet-28-small.py]: Dirk Xu 徐彬彬
Pursuing a Master's Degree 
Blog: https://freesouls.github.io
QQ: 765611336
Email: delcanxu@126.com or declanxu@gmail.com
Major: Computer Science & Technology @Zhejiang University
Lab: State Key Lab of CAD&CG @Zhejiang University
在2016年01月10日 18:36，Bing Xu 写道:

In example/image-classification/symbol_resnet-28-small.py:

> +        return act
> +
> +def residual_net(data, n):
> +    #fisrt 2n layers
> +    for i in xrange(0, n):
> +        data = residual_factory(data=data, num_filter=16, dim_match=True)
> +    
> +    #second 2n layers
> +    for i in xrange(0, n):
> +        if i==0:
> +            data = residual_factory(data=data, num_filter=32, dim_match=False)
> +        else:
> +            data = residual_factory(data=data, num_filter=32, dim_match=True)
> +    
> +    #third 2n layers
> +    for i in xrange(0, n):


Change to range for python3

**Comment by freesouls** [comment]
I will change it in hours for I am outside the moment. :)

Thanks

**Comment by freesouls** [comment]
@antinucleon change range for python3, and combine multiple commits into one

**Comment by shuokay** [comment]
almost the same to #1041 ~

**Comment by freesouls** [comment]
@shuokay not exactly, there are at least 4 differences 1) batch_normalization is applied in different places after second convolution, 2) shortcuts connection is also a little bit different, 3) initialization of the parameters and 4) train policy(lr, epoch etc.) 

before I got this version of network, I have trained several versions(mainly differ in where to add the batch_normalization, train policy), I have tried your version, got similar results as yours.

**Comment by shuokay** [comment]
@freesouls I have double checked the code and you are right~
